### ISSUE TITLE
Python 3.12 compatibility

### ISSUE THREAD

**Issue by braingram** [asdf-format/asdf#1634]
I was able to get the test suite to run on 3.12 (without fsspec which fails due to aiohttp failing to install).

https://github.com/asdf-format/asdf/actions/runs/5927002305/job/16069559190?pr=1633#step:10:183

I see no errors but 3 of the same warning:
```
DeprecationWarning: datetime.utcfromtimestamp() is deprecated and scheduled for removal in a future version. Use timezone-aware objects to represent datetimes in UTC: datetime.fromtimestamp(timestamp, datetime.UTC).
```

**Comment by braingram** [comment]
The changes here: https://github.com/asdf-format/asdf/pull/1633/commits/cc66fbfa4551df9c34ff0227b59cd831cf8d2399 allowed the test suite to pass on 3.12 with no warnings. Once the fsspec is updated and the other deps are cleaned up (as in #1633) we can claim 3.12 support.

**Comment by braingram** [comment]
Fixed in #1641
